### PR TITLE
Support limiting scope to one network or layer - full version

### PR DESCRIPTION
**Pull request by csillag**
It's now possible to specify a given network or layer in the .env file, and when given, the scope of the whole
explorer will be limited to that specific network, or even a specific layer within that network.
No data external to the required scope will be shown. The network (and the layer) will not be part of the URLs.

You can specify just the network, or just the layer, or both. This all results in different behavior.

   * * *

This is part of a larger effort to white-label the Explorer, so that it can be deployed for specific applications.

### CODE REVIEW

**Comment by csillag** [comment]
@donouwens  says: 

> All the limiting cases make 100% sense to me on desktop and mobile.
> [...]
> I’m happy to give a go 🟢  and 👍🏻  what you have right now @csillag

File: src/app/components/LayerPicker/index.tsx [hunk 1]
```diff
@@ -77,8 +77,8 @@ type LayerPickerContentProps = Omit<LayerPickerProps, 'open'>
 
 enum LayerPickerTabletStep {
   Network,
-  ParaTime,
```
Comment (lubej) [inline src/app/components/LayerPicker/index.tsx:80]: Doesn't ParaTime make more sense than Layer, and the LayerPicker should be renamed to ParaTime

File: src/app/components/PageLayout/Header.tsx [hunk 2]
```diff
@@ -49,7 +66,7 @@ export const Header: FC = () => {
               showText={!scrollTrigger && !isMobile}
             />
           </Grid>
-          {scope && (
+          {scope && needsScopeSelector() && (
```
Comment (lubej) [inline src/app/components/PageLayout/Header.tsx:69]: I would avoid doing function calls, isn't this static? In a sense that it wouldn't change on each re-render?

File: src/app/components/PageLayout/NetworkButton.tsx [hunk 3]
```diff
@@ -115,11 +119,14 @@ export const StyledMobileNetworkButton = styled(Button)(({ theme }) => ({
 
 export const MobileNetworkButton: FC<NetworkButtonProps> = ({ isOutOfDate, layer, network, onClick }) => {
   const { t } = useTranslation()
-  const labels = getLayerLabels(t)
+
+  const label = hasFixedLayer() // If we are fixed to a layer,
+    ? getNetworkNames(t)[network] // let's show the name of the network,
+    : getLayerLabels(t)[layer] // otherwise, the name of the layer.
```
Comment (lubej) [inline src/app/components/PageLayout/NetworkButton.tsx:125]: This could be some utility, it is duplicted twice in this file.

File: src/app/components/PageLayout/NetworkSelector.tsx [hunk 4]
```diff
@@ -84,7 +84,7 @@ const NetworkSelectorView: FC<NetworkSelectorViewProps> = ({ isOutOfDate, layer,
       {!isMobile && (
         <NetworkButton isOutOfDate={isOutOfDate} layer={layer} network={network} onClick={handleDrawerOpen} />
       )}
-      {!isTablet && network !== Network.mainnet && (
+      {!hasFixedLayer() && !isTablet && network !== Network.mainnet && (
```
Comment (lubej) [inline src/app/components/PageLayout/NetworkSelector.tsx:87]: Avoid function calls, this probably stays the same for the whole session. So the returned value, can be static.

File: src/app/components/ThemeByNetwork/index.tsx [hunk 5]
```diff
@@ -14,6 +15,10 @@ export const ThemeByNetwork: FC<{ network: Network; children: React.ReactNode }>
...
 )
 
-export const withDefaultTheme = (node: ReactNode) => (
+export const withMainnetTheme = (node: ReactNode) => (
   <ThemeByNetwork network={Network.mainnet}>{node}</ThemeByNetwork>
 )
+
+export const withDefaultTheme = (node: ReactNode) => (
+  <ThemeByNetwork network={getFixedNetwork() ?? Network.mainnet}>{node}</ThemeByNetwork>
+)
```
Comment (lubej) [inline src/app/components/ThemeByNetwork/index.tsx:24]: Should we change theme? Unless with fixed ParaTime only, you can have both mainnet and tesnet?

File: src/app/hooks/useSearchQueryNetworkParam.ts [hunk 6]
```diff
@@ -12,6 +12,12 @@ export const useSearchQueryNetworkParam = (): {
   setNetwork: (network: Network) => void
 } => {
   const [searchParams, setSearchParams] = useSearchParams()
+  if (hasFixedNetwork()) {
```
Comment (lubej) [inline src/app/hooks/useSearchQueryNetworkParam.ts:15]: Think that this could be static.

File: src/app/pages/HomePage/Graph/ParaTimeSelector/index.tsx [hunk 7]
```diff
@@ -284,7 +285,7 @@ const ParaTimeSelectorCmp: FC<ParaTimeSelectorProps> = ({
             <HelpScreen setParaTimeStep={setStep} />
           )}
         </ParaTimeSelectorGlobe>
-        {step === ParaTimeSelectorStep.Explore && (
+        {!hasFixedNetwork() && step === ParaTimeSelectorStep.Explore && (
```
Comment (lubej) [inline src/app/pages/HomePage/Graph/ParaTimeSelector/index.tsx:288]: Showing graph with fixed network/paratime, makes no sense. Homepage should be different in that case, or directly navigate to dashbaord.

File: src/app/pages/SearchResultsPage/GlobalSearchResultsView.tsx [hunk 8]
```diff
@@ -30,6 +30,25 @@ export const GlobalSearchResultsView: FC<{
 
   const themes = getThemesForNetworks()
   const networkNames = getNetworkNames(t)
+
+  if (hasFixedNetwork()) {
```
Comment (lubej) [inline src/app/pages/SearchResultsPage/GlobalSearchResultsView.tsx:34]: Isn't there a better way to filter search results? Does it makes queries to all the networks/layers still, in case of fixed network/paratime?

File: src/app/utils/route-utils.ts [hunk 9]
```diff
@@ -7,6 +7,32 @@ import { Network } from '../../types/network'
 import { SearchScope } from '../../types/searchScope'
 import { isStableDeploy } from '../../config'
 import { getSearchTermFromRequest } from '../components/Search/search-utils'
+import { isLayerNotHidden } from '../../types/layers'
+
+export const isTesting = () => process.env.NODE_ENV === 'test'
```
Comment (lubej) [inline src/app/utils/route-utils.ts:12]: This should be in env utils. Not here

**Review by lubej — COMMENTED** [review]
Didn't check all the code yet, partial review.

File: src/app/components/LayerPicker/index.tsx (hunk 1 above)
Comment (csillag) [reply src/app/components/LayerPicker/index.tsx:80]: Why? We have consensus, and then we have the paratimes. Here, we can select both, not only the paratimes. The  term that covers both is layer, right?

File: src/app/hooks/useSearchQueryNetworkParam.ts (hunk 6 above)
Comment (csillag) [reply src/app/hooks/useSearchQueryNetworkParam.ts]: Fixed

File: src/app/components/PageLayout/NetworkSelector.tsx (hunk 4 above)
Comment (csillag) [reply src/app/components/PageLayout/NetworkSelector.tsx]: Fixed

File: src/app/pages/HomePage/Graph/ParaTimeSelector/index.tsx (hunk 7 above)
Comment (csillag) [reply src/app/pages/HomePage/Graph/ParaTimeSelector/index.tsx]: Yes, homepage is different.

We _don't_ show the graph when we have a fixed layer. We still show it when we have a fixed network, just without the network switch.

You can check out the demos here: 
- [limited to a network](https://e97738f8.oasis-explorer.pages.dev/)
- [limited to a layer](https://84b385ed.oasis-explorer.pages.dev/)
-  [limited to both network and layer](https://4700fb51.oasis-explorer.pages.dev/)

File: src/app/utils/route-utils.ts (hunk 9 above)
Comment (csillag) [reply src/app/utils/route-utils.ts]: OK, moved to `utils/env.ts`

File: src/app/components/PageLayout/NetworkButton.tsx (hunk 3 above)
Comment (csillag) [reply src/app/components/PageLayout/NetworkButton.tsx]: OK, created the `getNetworkButtonLabel()` function.

File: src/app/pages/HomePage/Graph/ParaTimeSelector/index.tsx (hunk 7 above)
Comment (csillag) [reply src/app/pages/HomePage/Graph/ParaTimeSelector/index.tsx]: So the only (limited) case when we still want to show the graph is when we are limited to a network, but layer selection is free. In that case, we can still show the selector graph, just hide the mainnet/testnet switch. And this hiding is the code change here. So I think we are good here.

File: src/app/pages/SearchResultsPage/GlobalSearchResultsView.tsx (hunk 8 above)
Comment (csillag) [reply src/app/pages/SearchResultsPage/GlobalSearchResultsView.tsx]: Yes, good point. We shouldn't even make the queries. Will fix.

Comment (csillag) [reply src/app/pages/SearchResultsPage/GlobalSearchResultsView.tsx]: So, looking at the code, we were already limiting the queries to the right network, but not the right layer. Now I have fixed that, too.

So we don't need to filter when displaying the results, so removed that. (But we still need custom code for displaying, because we don't even want to offer other networks, so that part remains.)

**Comment by csillag** [comment]
> Didn't check all the code yet, partial review.

Yeah, thanks, responded to / solved all comments, except one that needs more thinking.

File: src/app/components/LayerPicker/index.tsx (hunk 1 above)
Comment (lubej) [reply src/app/components/LayerPicker/index.tsx:80]: Ah you are right, I always forget Consensus :) Layer makes perfect sense.

File: src/app/utils/route-utils.ts [hunk 10]
```diff
@@ -7,6 +7,33 @@ import { Network } from '../../types/network'
...
+
+export const fixedNetwork = isTesting
+  ? undefined
+  : (process.env.REACT_APP_FIXED_NETWORK as Network | undefined)
+export const fixedLayer = isTesting ? undefined : (process.env.REACT_APP_FIXED_LAYER as Layer | undefined)
+
+export type ScopeFreedom =
+  | 'network' // We can select only the network
+  | 'layer' // We can select only the layer
+  | 'both' // We can select both network and layer
```
Comment (lubej) [inline src/app/utils/route-utils.ts:21]: Just a nit, but I would name it network-layer or something similar.

File: src/types/layers.ts [hunk 11]
```diff
@@ -33,5 +33,6 @@ export const doesAnyOfTheseLayersSupportEncryptedTransactions = (layers: Layer[]
   uniq(layers).some(doesLayerSupportEncryptedTransactions)
 
 export const isLayerHidden = (layer: Layer): boolean => hiddenLayers.includes(layer)
+export const isLayerNotHidden = (layer: Layer): boolean => !hiddenLayers.includes(layer)
```
Comment (lubej) [inline src/types/layers.ts:36]: Not sure we need this? `!isLayerHidden(layer)`?

File: src/routes.tsx [hunk 12]
```diff
@@ -41,13 +50,79 @@ import { ProposalDetailsPage } from './app/pages/ProposalDetailsPage'
...
+ */
+const unlessFixedOnConsensus = (route: RouteObject): RouteObject[] => (isFixedOnConsensus ? [] : [route])
+
+/**
+ * Use the passed RouteObject if we have a fixed layer but no network
+ *
+ * Normally, this will just wrap the RouteObject into an array.
+ * Unless the condition is true, because in that case, and empty array will be returned.
+ */
+const ifFixedLayerButNoNetwork = (route: RouteObject): RouteObject[] =>
```
Comment (lubej) [inline src/routes.tsx:100]: I would skip `if` naming in a function.

File: src/routes.tsx [hunk 13]
```diff
@@ -41,13 +50,79 @@ import { ProposalDetailsPage } from './app/pages/ProposalDetailsPage'
...
+  if (layer) {
+    const network = RouteUtils.getEnabledNetworksForLayer(layer)[0]
+    if (network) {
+      to = RouteUtils.getDashboardRoute({ network, layer })
+    }
+  }
+  useEffect(() => {
+    if (to) navigate(to)
+  })
+  return <span>Wait...</span>
```
Comment (lubej) [inline src/routes.tsx:123]: Better loader here?

File: src/routes.tsx [hunk 14]
```diff
@@ -41,13 +50,79 @@ import { ProposalDetailsPage } from './app/pages/ProposalDetailsPage'
...
+/**
+ * In case of being restricted to a specific layer, jump to a network
+ *
+ * This should be rendered on the landing page,
+ * in order to redirect to a network, since we don't want the
+ * opening graph.
+ */
+const RedirectToNetwork: FC = () => {
+  const navigate = useNavigate()
+  const layer = fixedLayer
```
Comment (lubej) [inline src/routes.tsx:112]: This declaration is kind of useless.

File: src/app/components/ThemeByNetwork/index.tsx [hunk 15]
```diff
@@ -14,6 +15,10 @@ export const ThemeByNetwork: FC<{ network: Network; children: React.ReactNode }>
...
 )
 
-export const withDefaultTheme = (node: ReactNode) => (
+export const withMainnetTheme = (node: ReactNode) => (
   <ThemeByNetwork network={Network.mainnet}>{node}</ThemeByNetwork>
 )
+
+export const withDefaultTheme = (node: ReactNode) => (
+  <ThemeByNetwork network={fixedNetwork ?? Network.mainnet}>{node}</ThemeByNetwork>
+)
```
Comment (lubej) [inline src/app/components/ThemeByNetwork/index.tsx:24]: Still not sure why we need 2 Themes? And are we making sure, we are not accidentally including both?

File: src/app/components/PageLayout/Header.tsx [hunk 16]
```diff
@@ -8,6 +8,23 @@ import { NetworkSelector } from './NetworkSelector'
...
+    case 'network':
+      return RouteUtils.getEnabledNetworksForLayer(fixedLayer!).length > 1
+    case 'layer':
+      return RouteUtils.getVisibleLayersForNetwork(fixedNetwork!).length > 1
+    case 'both':
+      return RouteUtils.getEnabledScopes().length > 1
+    case 'none':
+      return false
+  }
+})() // We will define and call this function only once, and then store the result.
```
Comment (lubej) [inline src/app/components/PageLayout/Header.tsx:27]: Does this needs to be an IIFE? I would move this part to utils, and just call the function here, if it is needed only in Header. Or just get the static result from utils, since it probably won't change.

File: src/app/components/LayerPicker/index.tsx [hunk 17]
```diff
@@ -96,6 +96,9 @@ const LayerPickerContent: FC<LayerPickerContentProps> = ({ isOutOfDate, onClose,
   }
   const handleConfirm = () => onConfirm(selectedNetwork, selectedLayer)
 
+  // What kind of selection do we need?
+  const selectionMode = scopeFreedom
```
Comment (lubej) [inline src/app/components/LayerPicker/index.tsx:100]: Kinda useless declaration.

File: src/app/pages/HomePage/Graph/NetworkSelector/index.tsx [hunk 18]
```diff
@@ -119,12 +120,19 @@ const NetworkSelectorButton = forwardRef(
...
 }
 
-export const NetworkSelector: FC<NetworkSelectProps> = ({ network, setNetwork }) => {
+export const NetworkSelector: FC<NetworkSelectProps> = ({ network, setNetwork, forLayer }) => {
   const { t } = useTranslation()
 
-  const options = RouteUtils.getEnabledNetworks().map(network => ({
+  const networks = forLayer
+    ? RouteUtils.getEnabledNetworksForLayer(forLayer)
+    : RouteUtils.getEnabledNetworks()
```
Comment (lubej) [inline src/app/pages/HomePage/Graph/NetworkSelector/index.tsx:131]: I have seen this piece of code multiple times now, should be moved to utils I guess.

File: src/app/utils/route-utils.ts (hunk 10 above)
Comment (csillag) [reply src/app/utils/route-utils.ts]: OK. Done.

File: src/routes.tsx (hunk 14 above)
Comment (csillag) [reply src/routes.tsx]: Yeah it comes from the age when `fixedLayer()` was still a function. Now removed.

File: src/app/components/LayerPicker/index.tsx (hunk 17 above)
Comment (csillag) [reply src/app/components/LayerPicker/index.tsx]: Removed

File: src/types/layers.ts (hunk 11 above)
Comment (csillag) [reply src/types/layers.ts:36]: The difference between `!isLayerHidden(layer)` and `isLayerNotHidden()` is that we can use the latter as an arg for `.filter()`.

I find `layers.filter(isLayerNotHidden)` more readable than `layers.filter(layer => !isLayerHidden(layer))`

I think the extra declaration here is worth so that we can get an easier to read code elsewhere.

What do you think?

File: src/routes.tsx (hunk 13 above)
Comment (csillag) [reply src/routes.tsx]: Actually, we are going to be redirected right away, so we don't even need to render a loader. Returning null now, which seems to make it faster somehow.

File: src/routes.tsx (hunk 12 above)
Comment (csillag) [reply src/routes.tsx:100]: The semantics here is that we want to include this option only _if_ we have a fixed layer, but no (fixed) network.
So the function is not testing this condition, but doing something conditionally _if_ the condition is fulfilled.

What naming would you suggest?

Maybe `whenFixedLayerButNoNetwork()` ?

File: src/app/components/PageLayout/Header.tsx (hunk 16 above)
Comment (csillag) [reply src/app/components/PageLayout/Header.tsx]: Moved function to route-utils.

File: src/app/components/ThemeByNetwork/index.tsx (hunk 5 above)
Comment (csillag) [reply src/app/components/ThemeByNetwork/index.tsx]: This is also being discussed in another thread, so closing this thread.

File: src/app/components/ThemeByNetwork/index.tsx (hunk 15 above)
Comment (csillag) [reply src/app/components/ThemeByNetwork/index.tsx:24]: - `withMainnetTheme()` is only ever used for the external part of the `HomePage` (ie. opening graph). We must use it there, because the `HomePage` is not equipped to be displayed with the `TestTheme`.  (We would get black text on dark blue bg, etc.)
- However, the internal part of the `HomePage` (which shows the actual graph) will still get a theme corresponding to the network currently selected.
- The other wrapper, `withDefaultTheme()` is being used in other locations, when we can use any theme corresponding to the (default) network.

The reason we need 2 wrappers now, when we could do with 1 earlier, is that now  the default network might be testnet, while earlier it was always mainnet, and so when using `withDefaultTheme`, we might get the testnet theme as the default theme, which is not suitable for everywhere.

Please see the demos about how this works in practice.

File: src/app/pages/HomePage/Graph/NetworkSelector/index.tsx (hunk 18 above)
Comment (csillag) [reply src/app/pages/HomePage/Graph/NetworkSelector/index.tsx]: Done

File: src/types/layers.ts (hunk 11 above)
Comment (csillag) [reply src/types/layers.ts]: Removed it anyway.

File: src/routes.tsx (hunk 12 above)
Comment (lubej) [reply src/routes.tsx:100]: Yeah, maybe the one you suggested, or `getFixedLayerButNoNetwork`. I don't know, but `if` feels to me counterintuitive.

File: src/app/components/ThemeByNetwork/index.tsx (hunk 15 above)
Comment (lubej) [reply src/app/components/ThemeByNetwork/index.tsx:24]: Ok, still think that a simple if condition could fix this. And we could just use one theme. Or the HOC could accept additional params, to make the theme fixed, in the `HomePage` case. Either way, I guess it's fine.

File: src/app/components/PageLayout/Header.tsx [hunk 19]
```diff
@@ -49,7 +50,7 @@ export const Header: FC = () => {
               showText={!scrollTrigger && !isMobile}
             />
           </Grid>
-          {scope && (
+          {scope && isScopeSelectorNeeded(scope) && (
```
Comment (lubej) [inline src/app/components/PageLayout/Header.tsx:53]: Not a fan of function calls inside JSX.

File: src/app/utils/route-utils.ts [hunk 20]
```diff
@@ -7,6 +7,33 @@ import { Network } from '../../types/network'
 import { SearchScope } from '../../types/searchScope'
 import { isStableDeploy } from '../../config'
 import { getSearchTermFromRequest } from '../components/Search/search-utils'
+import { isLayerHidden } from '../../types/layers'
+import { isTesting } from './env'
+
+export const fixedNetwork = isTesting
+  ? undefined
+  : (process.env.REACT_APP_FIXED_NETWORK as Network | undefined)
+export const fixedLayer = isTesting ? undefined : (process.env.REACT_APP_FIXED_LAYER as Layer | undefined)
```
Comment (lubej) [inline src/app/utils/route-utils.ts:16]: Why is this needed? Why would we want to unset fixed network/layer for testing only?

**Review by lubej — APPROVED** [review]
LGTM overall

File: src/app/components/PageLayout/Header.tsx (hunk 19 above)
Comment (csillag) [reply src/app/components/PageLayout/Header.tsx:53]: Would you prefer having it declared as a const, before the return command?

File: src/app/utils/route-utils.ts (hunk 20 above)
Comment (csillag) [reply src/app/utils/route-utils.ts:16]: Because we want to use simple routes and paths for the tests, and don't want to rewrite them for complicated, conditional logic. 

It's easiest to always run the tests against the basic configuration, independently of how we plan to deploy the app.

In fact, _not_ unsetting for playwright also breaks _those_ tests, which is still an unsolved issue. (How do we easily detect that we are running inside playwrighet?)

File: src/routes.tsx (hunk 12 above)
Comment (csillag) [reply src/routes.tsx]: > > Maybe `whenFixedLayerButNoNetwork()` ?

> Yeah, maybe the one you suggested, 

OK, renamed.

File: src/app/components/ThemeByNetwork/index.tsx (hunk 15 above)
Comment (csillag) [reply src/app/components/ThemeByNetwork/index.tsx]: >  Or the HOC could accept additional params, to make the theme fixed, in the HomePage case.

Did that now. How do you like it now?

File: src/app/components/PageLayout/Header.tsx (hunk 19 above)
Comment (csillag) [reply src/app/components/PageLayout/Header.tsx]: never mind, moved to a const before the JSX part.

File: .changelog/1242.feature.md [hunk 21]
```diff
@@ -0,0 +1 @@
+Support limiting scope to one network or layer
```
Comment (lukaw3d) [inline .changelog/1242.feature.md:1]: This seems hard to maintain. Did any other approaches have issues?

Comment (csillag) [reply .changelog/1242.feature.md:1]: > This seems hard to maintain.

Well, I guess we can simplify it a great deal if we give up having the shortest possible URLs.... lemme prepare and alternate PR with that version.

Comment (csillag) [reply .changelog/1242.feature.md:1]: How about #1269?

Comment (lukaw3d) [reply .changelog/1242.feature.md:1]: It's not a different approach tho

Some other approaches:
- Do we only need to support one of `pinning network | pinning layer | pinning network+layer` for PontusX? Does it matter? Are there issues supporting just one?
- Issues if we tried to unify "fixed" vs unfixed by moving enabled layers to .env.production?
- Can react router intercept and rewrite paths
- ..
